langchain4j-core/src/test/java/dev/langchain4j/model/chat/common/StreamingChatLanguageModelIT.java

```diff
@@ -0,0 +1,228 @@
+ *
+ * <dependency>
+ *     <groupId>org.mockito</groupId>
+ *     <artifactId>mockito-core</artifactId>
+ *     <scope>test</scope>
+ * </dependency>
+ *
+ * <dependency>
+ *     <groupId>org.mockito</groupId>
+ *     <artifactId>mockito-junit-jupiter</artifactId>
+ *     <scope>test</scope>
+ * </dependency>
+ */
+@TestInstance(PER_CLASS)
+public abstract class StreamingChatLanguageModelIT {
+
+    protected abstract List<StreamingChatLanguageModel> models();
+
+    @ParameterizedTest
+    @MethodSource("models")
+    void should_answer_simple_question(StreamingChatLanguageModel model) throws Exception {
+
+        // given
+        model = spy(model);
+
+        ChatRequest chatRequest = ChatRequest.builder()
+                .messages(UserMessage.from("What is the capital of Germany?"))
+                .build();
+
+        CompletableFuture<ChatResponse> futureChatResponse = new CompletableFuture<>();
+        StringBuffer tokenAccumulator = new StringBuffer();
+        AtomicInteger timesOnPartialResponseIsCalled = new AtomicInteger();
+        Set<Thread> threads = new CopyOnWriteArraySet<>();
+
+        // when
+        model.chat(chatRequest, new StreamingChatResponseHandler() {
+
+            @Override
+            public void onPartialResponse(String partialResponse) {
+                tokenAccumulator.append(partialResponse);
+                timesOnPartialResponseIsCalled.incrementAndGet();
+                threads.add(Thread.currentThread());
+            }
+
+            @Override
+            public void onCompleteResponse(ChatResponse chatResponse) {
+                futureChatResponse.complete(chatResponse);
+                threads.add(Thread.currentThread());
+            }
+
+            @Override
+            public void onError(Throwable error) {
+                threads.add(Thread.currentThread());
+                futureChatResponse.completeExceptionally(error);
+            }
+        });
+
+        // then
+        ChatResponse chatResponse = futureChatResponse.get(30, SECONDS);
+
+        AiMessage aiMessage = chatResponse.aiMessage();
+        assertThat(aiMessage.text()).containsIgnoringCase("Berlin");
+        assertThat(aiMessage.toolExecutionRequests()).isNull();
+
+        assertThat(tokenAccumulator.toString()).isEqualTo(aiMessage.text());
+
+        if (assertTokenUsage()) {
+            TokenUsage tokenUsage = chatResponse.tokenUsage();
+            assertThat(tokenUsage.inputTokenCount()).isGreaterThan(0);
+            assertThat(tokenUsage.outputTokenCount()).isGreaterThan(0);
+            assertThat(tokenUsage.totalTokenCount())
+                    .isEqualTo(tokenUsage.inputTokenCount() + tokenUsage.outputTokenCount());
+        }
+
+        if (assertFinishReason()) {
+            assertThat(chatResponse.finishReason()).isEqualTo(STOP);
+        }
+
+        assertThat(timesOnPartialResponseIsCalled.get()).isGreaterThan(1);
+
+        if (assertThreads()) {
+            assertThat(threads).hasSize(1);
```

Comment:
Yes, we need to make sure the behavior is consistent between model implementations. Or at least be able to know the differences (if they cannot be fixed) and document them. This test is here to highlight the differences.

Some users already had concerns about tools and callbacks being called on different threads. Also because threads were created in internal thread pools and did not have security context propagated between them. So they were not able to call some of their services from the tool. Also we need to make sure the correlation IDs (for observability) are passed correctly between all threads.